ZipArchive.m

```diff
@@ -543,6 +538,262 @@ -(NSDictionary *)UnzipFileToMemory
+ */
+-(NSData *)UnzipFileDataWithFileName:(NSString *)fileName
+{
+    int ret = unzGoToFirstFile( _unzFile );
+    unsigned char		buffer[4096] = {0};
+    if( ret!=UNZ_OK )
+    {
+        [self OutputErrorMessage:@"Failed"];
+    }
+    
+    const char* password = [_password cStringUsingEncoding:NSASCIIStringEncoding];
+    
+    NSData *resultData = nil;
+    
+    do{
+        @autoreleasepool {
+            if( [_password length]==0 )
+                ret = unzOpenCurrentFile( _unzFile );
+            else
+                ret = unzOpenCurrentFilePassword( _unzFile, password );
+            if( ret!=UNZ_OK )
+            {
+                [self OutputErrorMessage:@"Error occurs"];
+                break;
+            }
+            // reading data and write to file
+            int read ;
+            unz_file_info	fileInfo ={0};
+            ret = unzGetCurrentFileInfo(_unzFile, &fileInfo, NULL, 0, NULL, 0, NULL, 0);
+            if( ret!=UNZ_OK )
+            {
+                [self OutputErrorMessage:@"Error occurs while getting file info"];
+                unzCloseCurrentFile( _unzFile );
+                break;
+            }
+            char* filename = (char*) malloc( fileInfo.size_filename +1 );
+            unzGetCurrentFileInfo(_unzFile, &fileInfo, filename, fileInfo.size_filename + 1, NULL, 0, NULL, 0);
+            filename[fileInfo.size_filename] = '\0';
+            
+            // check if it contains directory
+            NSString * strPath = [NSString stringWithCString:filename encoding:self.stringEncoding];
+            free( filename );
+            if( [strPath rangeOfCharacterFromSet:[NSCharacterSet characterSetWithCharactersInString:@"/\\"]].location!=NSNotFound )
+            {// contains a path
+                strPath = [strPath stringByReplacingOccurrencesOfString:@"\\" withString:@"/"];
+            }
+            
+            if (![strPath isEqualToString:fileName])
+            {
+                if (ret == UNZ_OK) {
+                    ret = unzCloseCurrentFile( _unzFile );
+                    if (ret != UNZ_OK) {
+                        [self OutputErrorMessage:@"file was unzipped but failed crc check"];
+                    }
+                }
+                
+                if (ret == UNZ_OK) {
+                    ret = unzGoToNextFile( _unzFile );
+                }
+                continue;
+            }
+            
+            NSMutableData *fileMutableData = [NSMutableData data];
+            do
+            {
+                read = unzReadCurrentFile(_unzFile, buffer, 4096);
+                if (read >= 0)
+                {
+                    if (read != 0)
+                    {
+                        [fileMutableData appendBytes:buffer length:read];
+                    }
+                }
+                else // if (read < 0)
+                {
+                    [self OutputErrorMessage:@"Failed to read zip file"];
+                }
+            } while (read > 0);
```

Comment:
This loop won't exit correctly if the number of read bytes is zero, no?

Reply:
If read bytes is zero or negative value, the do-while loop breaks. If positive value, data appended into fileMutableData.